### PR TITLE
planner: Add explain info for ClusterLogTableExtractor

### PR DESCRIPTION
**Pull request by Yisaer**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/15379

Problem Summary:

UCP: Add explain info for ClusterLogTableExtractor
1. planner/core: memtable_predicate_extractor.go
2. *: Add explain info for ClusterLogTableExtractor

Here is an example I query it locally：

```sh
mysql> desc select * from information_schema.cluster_log where time >= '2019-12-23 16:10:13' and time <= '2019-12-23 16:30:13';
+----------------+----------+------+-----------------------------------------------------------------------------------------------------------------------------+
| id             | estRows  | task | operator info                                                                                                               |
+----------------+----------+------+-----------------------------------------------------------------------------------------------------------------------------+
| MemTableScan_5 | 10000.00 | root | table:CLUSTER_LOG, Start_time:2019-12-23 16:10:13, end_time:2019-12-23 16:30:13|
+----------------+----------+------+-----------------------------------------------------------------------------------------------------------------------------+
```

### Check List 

Tests 

- Unit test

Side effects

- No

### CODE REVIEW

File: planner/core/memtable_predicate_extractor.go [hunk 1]
```diff
@@ -606,7 +606,16 @@ func (e *ClusterLogTableExtractor) Extract(
...
+	if e.SkipRequest {
+		return "skip_request: true"
+	}
+	startTime, endTime := e.StartTime, e.EndTime
+	st := time.Unix(0, startTime*1e6)
+	et := time.Unix(0, endTime*1e6)
+	return fmt.Sprintf("Start_time:%v, end_time:%v",
+		st.In(p.ctx.GetSessionVars().StmtCtx.TimeZone).Format(MetricTableTimeFormat),
+		et.In(p.ctx.GetSessionVars().StmtCtx.TimeZone).Format(MetricTableTimeFormat),
+	)
```
Comment (crazycs520) [inline planner/core/memtable_predicate_extractor.go]: Please add test for this, you can refer `TestExplainMetricTable` function.

File: planner/core/memtable_predicate_extractor.go [hunk 2]
```diff
@@ -606,7 +606,16 @@ func (e *ClusterLogTableExtractor) Extract(
...
 
 func (e *ClusterLogTableExtractor) explainInfo(p *PhysicalMemTable) string {
-	return ""
+	if e.SkipRequest {
+		return "skip_request: true"
+	}
+	startTime, endTime := e.StartTime, e.EndTime
+	st := time.Unix(0, startTime*1e6)
+	et := time.Unix(0, endTime*1e6)
+	return fmt.Sprintf("Start_time:%v, end_time:%v",
```
Comment (crazycs520) [inline planner/core/memtable_predicate_extractor.go]: Need more information, like `NodeTypes`, `Instances`, `Patterns`, `LogLevels`

File: planner/core/memtable_predicate_extractor.go (hunk 1 above)
Comment (Yisaer) [reply planner/core/memtable_predicate_extractor.go]: updated.

File: planner/core/memtable_predicate_extractor.go (hunk 2 above)
Comment (Yisaer) [reply planner/core/memtable_predicate_extractor.go]: updated.

File: planner/core/util.go [hunk 3]
```diff
@@ -231,3 +233,15 @@ func GetStatsInfo(i interface{}) map[string]uint64 {
 	statsInfos = CollectPlanStatsVersion(physicalPlan, statsInfos)
 	return statsInfos
 }
+
+// ExtractStringFromStringSet helps extract string info from set.StringSet
+func ExtractStringFromStringSet(set set.StringSet) string {
```
Comment (crazycs520) [inline planner/core/util.go]: No need to export this function?

File: planner/core/util.go [hunk 4]
```diff
@@ -231,3 +233,15 @@ func GetStatsInfo(i interface{}) map[string]uint64 {
...
+	r := ""
+	if len(set) < 1 {
+		return r
+	}
+	for k := range set {
```
Comment (crazycs520) [inline planner/core/util.go]: Please add a sort here to make the result stable.

currently

```sql
information_schema> desc select * from `CLUSTER_LOG` where level in ('warn','error')
+----------------+----------+------+-------------------------------------------------------------------------------------------------------------------------------------------+
| id             | estRows  | task | operator info                                                                                                                             |
+----------------+----------+------+-------------------------------------------------------------------------------------------------------------------------------------------+
| MemTableScan_5 | 10000.00 | root | table:CLUSTER_LOG, Start_time:1970-01-01 08:00:00, end_time:1970-01-01 07:59:59.999, node_types:[], instances:[], log_levels:[warn,error] |
+----------------+----------+------+-------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set
Time: 0.007s
information_schema> desc select * from `CLUSTER_LOG` where level in ('warn','error')
+----------------+----------+------+-------------------------------------------------------------------------------------------------------------------------------------------+
| id             | estRows  | task | operator info                                                                                                                             |
+----------------+----------+------+-------------------------------------------------------------------------------------------------------------------------------------------+
| MemTableScan_5 | 10000.00 | root | table:CLUSTER_LOG, Start_time:1970-01-01 08:00:00, end_time:1970-01-01 07:59:59.999, node_types:[], instances:[], log_levels:[error,warn] |
+----------------+----------+------+-------------------------------------------------------------------------------------------------------------------------------------------+
```

File: planner/core/util.go [hunk 5]
```diff
@@ -231,3 +233,15 @@ func GetStatsInfo(i interface{}) map[string]uint64 {
...
+		r = fmt.Sprintf("%s,%s", r, k)
```
Comment (crazycs520) [inline planner/core/util.go]: How about using ` bytes.Buffer` instead of `fmt.Sprintf`?

File: planner/core/util.go (hunk 4 above)
Comment (Yisaer) [reply planner/core/util.go]: I wonder is there any existed tool or function in tidb to sort the `set.String` or i need to build an extra `type` to implement the `sort`?

File: planner/core/memtable_predicate_extractor.go [hunk 6]
```diff
@@ -606,7 +606,17 @@ func (e *ClusterLogTableExtractor) Extract(
...
+	if e.SkipRequest {
+		return "skip_request: true"
+	}
+	startTime, endTime := e.StartTime, e.EndTime
+	st := time.Unix(0, startTime*1e6)
+	et := time.Unix(0, endTime*1e6)
+	return fmt.Sprintf("Start_time:%v, end_time:%v, node_types:[%s], instances:[%s], log_levels:[%s]",
+		st.In(p.ctx.GetSessionVars().StmtCtx.TimeZone).Format(MetricTableTimeFormat),
+		et.In(p.ctx.GetSessionVars().StmtCtx.TimeZone).Format(MetricTableTimeFormat),
+		ExtractStringFromStringSet(e.NodeTypes), ExtractStringFromStringSet(e.Instances), ExtractStringFromStringSet(e.LogLevels),
```
Comment (crazycs520) [inline planner/core/memtable_predicate_extractor.go]: Please add `len(e.NodeTypes) > 0` check.

It's hard to see the difference of `level`at below:

```sql
information_schema> desc select * from `CLUSTER_LOG`;
+----------------+----------+------+-----------------------------------------------------------------------------------------------------------------------------+
| id             | estRows  | task | operator info                                                                                                               |
+----------------+----------+------+-----------------------------------------------------------------------------------------------------------------------------+
| MemTableScan_4 | 10000.00 | root | table:CLUSTER_LOG, Start_time:1970-01-01 08:00:00, end_time:1970-01-01 08:00:00, node_types:[], instances:[], log_levels:[] |
+----------------+----------+------+-----------------------------------------------------------------------------------------------------------------------------+
1 row in set
Time: 0.011s
information_schema> desc select * from `CLUSTER_LOG` where level = ''
+----------------+----------+------+---------------------------------------------------------------------------------------------------------------------------------+
| id             | estRows  | task | operator info                                                                                                                   |
+----------------+----------+------+---------------------------------------------------------------------------------------------------------------------------------+
| MemTableScan_5 | 10000.00 | root | table:CLUSTER_LOG, Start_time:1970-01-01 08:00:00, end_time:1970-01-01 07:59:59.999, node_types:[], instances:[], log_levels:[] |
+----------------+----------+------+---------------------------------------------------------------------------------------------------------------------------------+

```

File: planner/core/util.go (hunk 4 above)
Comment (Yisaer) [reply planner/core/util.go]: updated.

File: planner/core/util.go (hunk 3 above)
Comment (Yisaer) [reply planner/core/util.go]: updated.

File: planner/core/util.go (hunk 5 above)
Comment (Yisaer) [reply planner/core/util.go]: updated.

File: planner/core/memtable_predicate_extractor.go (hunk 6 above)
Comment (Yisaer) [reply planner/core/memtable_predicate_extractor.go]: updated.

File: planner/core/util.go [hunk 7]
```diff
@@ -231,3 +235,33 @@ func GetStatsInfo(i interface{}) map[string]uint64 {
...
+	r := new(bytes.Buffer)
+	if len(set) < 1 {
+		return r.String()
+	}
+	l := make([]string, 0, len(set))
+	for k := range set {
+		l = append(l, k)
+	}
+	sl := StringList(l)
+	sort.Sort(sl)
```
Comment (crazycs520) [inline planner/core/util.go]: You can use `	sort.Strings()`, and remove line #238 ~ 250.

Comment (Yisaer) [reply planner/core/util.go]: updated.

File: planner/core/memtable_predicate_extractor.go [hunk 8]
```diff
@@ -606,7 +606,56 @@ func (e *ClusterLogTableExtractor) Extract(
 }
 
 func (e *ClusterLogTableExtractor) explainInfo(p *PhysicalMemTable) string {
-	return ""
+	if e.SkipRequest {
+		return "skip_request: true"
+	}
+	r := new(bytes.Buffer)
+	st, et := e.GetTimeRange(true)
```
Comment (crazycs520) [inline planner/core/memtable_predicate_extractor.go]: Should use `GetTimeRange(false)`.

File: planner/core/util.go [hunk 9]
```diff
@@ -231,3 +235,20 @@ func GetStatsInfo(i interface{}) map[string]uint64 {
...
 }
+
+// extractStringFromStringSet helps extract string info from set.StringSet
+func extractStringFromStringSet(set set.StringSet) string {
+	r := new(bytes.Buffer)
+	if len(set) < 1 {
+		return r.String()
+	}
+	l := make([]string, 0, len(set))
+	for k := range set {
```
Comment (crazycs520) [inline planner/core/util.go]: prefer 
```go
    for i, k := range l {
		if i > 0 {
			r.WriteByte(',')
		}
		r.WriteByte('"')
		r.WriteString(k)
		r.WriteByte('"')
	}
	return r.String()
```

Comment (Yisaer) [reply planner/core/util.go]: updated.

File: planner/core/memtable_predicate_extractor.go (hunk 8 above)
Comment (Yisaer) [reply planner/core/memtable_predicate_extractor.go]: updated.

**Comment by crazycs520** [comment]
@lonng PTAL

File: planner/core/memtable_predicate_extractor.go [hunk 10]
```diff
@@ -606,7 +606,56 @@ func (e *ClusterLogTableExtractor) Extract(
...
+	if st > 0 {
+		st := time.Unix(0, st*1e6)
+		r.WriteString(fmt.Sprintf("start_time:%v, ", st.In(p.ctx.GetSessionVars().StmtCtx.TimeZone).Format(MetricTableTimeFormat)))
+	}
+	if et < math.MaxInt64 {
+		et := time.Unix(0, et*1e6)
+		r.WriteString(fmt.Sprintf("end_time:%v, ", et.In(p.ctx.GetSessionVars().StmtCtx.TimeZone).Format(MetricTableTimeFormat)))
+	}
+	if len(e.NodeTypes) > 0 {
+		r.WriteString(fmt.Sprintf("node_types:[%s], ", extractStringFromStringSet(e.Instances)))
```
Comment (lonng) [inline planner/core/memtable_predicate_extractor.go]:
```suggestion
		r.WriteString(fmt.Sprintf("node_types:[%s], ", extractStringFromStringSet(e. NodeTypes)))
```?

File: planner/core/util.go [hunk 11]
```diff
@@ -231,3 +235,25 @@ func GetStatsInfo(i interface{}) map[string]uint64 {
...
+		l = append(l, k)
+	}
+	sort.Strings(l)
+	for i, k := range l {
```
Comment (lonng) [inline planner/core/util.go]: return fmt.Sprintf(`"%s"`,strings.Join(l, `","`))

File: executor/explainfor_test.go [hunk 12]
```diff
@@ -112,4 +112,8 @@ func (s *testSuite) TestExplainMetricTable(c *C) {
 		"MemTableScan_5 10000.00 root table:tidb_query_duration, PromQL:histogram_quantile(0.9, sum(rate(tidb_server_handle_query_duration_seconds_bucket{}[60s])) by (le,sql_type,instance)), start_time:2019-12-23 16:10:13, end_time:2019-12-23 16:30:13, step:1m0s"))
 	tk.MustQuery(fmt.Sprintf("desc select * from METRICS_SCHEMA.up where time >= '2019-12-23 16:10:13' and time <= '2019-12-23 16:30:13' ")).Check(testkit.Rows(
 		"MemTableScan_5 10000.00 root table:up, PromQL:up{}, start_time:2019-12-23 16:10:13, end_time:2019-12-23 16:30:13, step:1m0s"))
+	tk.MustQuery("desc select * from information_schema.cluster_log where time >= '2019-12-23 16:10:13' and time <= '2019-12-23 16:30:13'").Check(testkit.Rows(
+		"MemTableScan_5 10000.00 root table:CLUSTER_LOG, start_time:2019-12-23 16:10:13, end_time:2019-12-23 16:30:13"))
+	tk.MustQuery("desc select * from information_schema.cluster_log where level in ('warn','error') and time >= '2019-12-23 16:10:13' and time <= '2019-12-23 16:30:13'").Check(testkit.Rows(
```
Comment (lonng) [inline executor/explainfor_test.go]: Add some test cases for node types.

Comment (Yisaer) [reply executor/explainfor_test.go]: updated

File: planner/core/util.go (hunk 11 above)
Comment (Yisaer) [reply planner/core/util.go]: updated.

**Comment by Yisaer** [comment]
@crazycs520 @lonng  PTAL

**Pull request merged**